File: apps/api/src/routes/health.ts
import { Router, Request, Response } from 'express';
import { checkDatabaseHealth } from '../lib/db';
import { checkRedisHealth } from '../lib/redis';
import { env } from '../config/env';

const router = Router();

interface HealthStatus {
    status: 'ok' | 'degraded' | 'unhealthy';
    service: string;
    version: string;
    timestamp: string;
    uptime: number;
    checks: {
        database: boolean;
        redis: boolean;
    };
}

/**
 * Health Check Endpoint
 * 
 * GET /health
 * 
 * Returns service health status including:
 * - Database connectivity
 * - Redis connectivity
 * - Service uptime
 */
router.get('/', async (_req: Request, res: Response) => {
    const startTime = process.hrtime();

    const [dbHealthy, redisHealthy] = await Promise.all([
        checkDatabaseHealth().catch(() => false),
        checkRedisHealth().catch(() => false),
    ]);

    const allHealthy = dbHealthy && redisHealthy;

    const healthStatus: HealthStatus = {
        status: allHealthy ? 'ok' : dbHealthy || redisHealthy ? 'degraded' : 'unhealthy',
        service: env.SERVICE_NAME,
        version: process.env['npm_package_version'] ?? '0.1.0',
        timestamp: new Date().toISOString(),
        uptime: process.uptime(),
        checks: {
            database: dbHealthy,
            redis: redisHealthy,
        },
    };

    const statusCode = allHealthy ? 200 : 503;
    res.status(statusCode).json(healthStatus);
});

/**
 * Liveness Probe
 * 
 * GET /health/live
 * 
 * Simple check that the service is running.
 * Used by Kubernetes liveness probes.
 */
router.get('/live', (_req: Request, res: Response) => {
    res.status(200).json({ status: 'ok' });
});

/**
 * Readiness Probe
 * 
 * GET /health/ready
 * 
 * Checks if the service is ready to accept traffic.
 * Used by Kubernetes readiness probes.
 */
router.get('/ready', async (_req: Request, res: Response) => {
    const [dbHealthy, redisHealthy] = await Promise.all([
        checkDatabaseHealth().catch(() => false),
        checkRedisHealth().catch(() => false),
    ]);

    if (dbHealthy && redisHealthy) {
        res.status(200).json({ status: 'ready' });
    } else {
        res.status(503).json({
            status: 'not_ready',
            checks: { database: dbHealthy, redis: redisHealthy },
        });
    }
});

export default router;

File: services/ai-engine/app/routers/health.py
"""
TransLogistics AI Engine - Health Check Router

Provides health and readiness endpoints for container orchestration.
"""

import time
from typing import Literal

from fastapi import APIRouter, Response
from pydantic import BaseModel

from app.config import get_settings
from app.utils.logger import logger


router = APIRouter(prefix="/health", tags=["Health"])

# Track start time for uptime calculation
START_TIME = time.time()


class HealthResponse(BaseModel):
    """Health check response model."""
    
    status: Literal["ok", "degraded", "unhealthy"]
    service: str
    version: str
    uptime_seconds: float
    model_version: str


class LivenessResponse(BaseModel):
    """Liveness probe response."""
    
    status: Literal["ok"]


class ReadinessResponse(BaseModel):
    """Readiness probe response."""
    
    status: Literal["ready", "not_ready"]
    checks: dict[str, bool]


@router.get("", response_model=HealthResponse)
async def health_check() -> HealthResponse:
    """
    Health Check Endpoint
    
    Returns comprehensive health status including:
    - Service status
    - Version information
    - Uptime
    - Model version
    """
    settings = get_settings()
    
    return HealthResponse(
        status="ok",
        service=settings.APP_NAME,
        version=settings.APP_VERSION,
        uptime_seconds=round(time.time() - START_TIME, 2),
        model_version=settings.VOLUMESCAN_MODEL_VERSION,
    )


@router.get("/live", response_model=LivenessResponse)
async def liveness_probe() -> LivenessResponse:
    """
    Liveness Probe
    
    Simple check that the service is running.
    Used by Kubernetes liveness probes.
    """
    return LivenessResponse(status="ok")


@router.get("/ready", response_model=ReadinessResponse)
async def readiness_probe(response: Response) -> ReadinessResponse:
    """
    Readiness Probe
    
    Checks if the service is ready to accept traffic.
    Used by Kubernetes readiness probes.
    
    For AI Engine, we check:
    - Model availability (placeholder for now)
    """
    # Placeholder checks - will be expanded when model is integrated
    checks = {
        "model_loaded": True,  # Placeholder
    }
    
    all_ready = all(checks.values())
    
    if not all_ready:
        response.status_code = 503
        logger.warning("Readiness check failed", checks=checks)
    
    return ReadinessResponse(
        status="ready" if all_ready else "not_ready",
        checks=checks,
    )
